tests/unit_tests/CMakeLists.txt

```diff
@@ -20,4 +20,5 @@ endif()
 # Set up all of the "device" tests.
 if( DETRAY_BUILD_CUDA )
    add_subdirectory( cuda )
-endif()
+   add_subdirectory( vecpar )
```

Comment:
:thinking: If in the test we end up hardcoding CUDA support (through its memory management), then things don't seem quite as flexible as they should be, no?

It may be worth demonstrating already here how you propose making the code independent of the concrete backend that we want to use in a given source file. Since I'm not really seeing that yet. :frowning:

Clearly you should separate the creation of the memory resource(s) used by the test from the rest of the test code somehow. I could imagine either a factory pattern, or possibly simply just building the test with different implementations of a common

```c++
namespace vecpar_test {
vecmem::memory_resource& get_memory_resource();
}
```

function, depending on what backend the algorithmic part of the test is being compiled for.

Reply:
@krasznaa I'm afraid that I didn't explain it too well nor implemented too clear. I will try to do this now: there are (potentially) 2 types of memory resources (host and managed) and two types of architectures (host and device). 
The algorithm itself implemented in `.hpp` file is the same regardless of the memory resource or the arch that is compiled for. What might change, it is the test, so we could have the following scenarios:

1. the data is allocated in managed memory (so the cuda toolkit is available on the class path)
 - the test can be compiled for CPU by *any* C++ compiler -> the result is `detray_test_{plugin}_vecpar_cpu`
 - the test can be compiled for NVIDIA GPU by clang compiler *only* (because gcc cannot compile CUDA code) -> the result is `detray_test_{plugin}_vecpar_gpu`
 
2. the data is allocated in host memory
    - the test can be compiled for CPU by *any* C++ compiler -> the result is `detray_test_{plugin}_vecpar_cpu_host_mem`
    - *HERE IS THE MISSING PIECE*  the test can be compiled for NVIDIA GPU by clang compiler *only* if, and only if, the data that needs to be transferred to-from device `is_default_constructible`. vecpar uses `vecmem::copy` to do this behind the scene but the last copy device-host fails. Maybe I'm not using vecmem correctly [here](https://github.com/wr-hamburg/vecpar/blob/432d32aee1ae082609060dc074fe6f2558b32417/backend/cuda/include/vecpar/cuda/detail/host_memory.hpp#L82), I would really benefit out of your feedback on this.
Due to the missing piece and in order to ensure that the code builds without it, I commented out this scenario [here](https://github.com/georgi-mania/detray/blob/c3e76a48d32e4c2261faaf958f5624ca136679e0/tests/unit_tests/vecpar/CMakeLists.txt#L63).

I have re-organized the implementation for clarity.